### PR TITLE
SDIT-1649 Expose all historical identifiers to prisoner search

### PR DESCRIPTION
**Pull request by mikehalmamoj**

### CODE REVIEW

File: src/main/java/uk/gov/justice/hmpps/prison/service/PrisonerSearchService.kt [hunk 1]
```diff
@@ -46,7 +47,8 @@ class PrisonerSearchService(
           csra = it.csra,
           categoryCode = it.categoryCode,
           inOutStatus = it.inOutStatus,
-          identifiers = it.identifiers?.sortedBy { it.whenCreated },
+          identifiers = it.identifiers?.filter { id -> id.offenderId == it.offenderId }?.sortedBy { it.whenCreated },
```
Comment (mikehalmamoj) [inline src/main/java/uk/gov/justice/hmpps/prison/service/PrisonerSearchService.kt:50]: Continue to provide identifiers for the active booking for now.

File: src/main/java/uk/gov/justice/hmpps/prison/service/PrisonerSearchService.kt [hunk 2]
```diff
@@ -46,7 +47,8 @@ class PrisonerSearchService(
...
+          allIdentifiers = it.identifiers?.sortedBy { it.whenCreated },
```
Comment (mikehalmamoj) [inline src/main/java/uk/gov/justice/hmpps/prison/service/PrisonerSearchService.kt:51]: But `identifiers` will be replaced by `allIdentifiers` so we can see identifires for aliases.

File: src/main/java/uk/gov/justice/hmpps/prison/api/model/Alias.java [hunk 3]
```diff
@@ -58,4 +58,8 @@ public class Alias {
     @NotNull
     @Schema(description = "Date of creation", example = "2019-02-15", requiredMode = RequiredMode.NOT_REQUIRED)
     private LocalDate createDate;
+
+    @NotNull
+    @Schema(description = "Offender ID", example = "543548")
+    private Long offenderId;
```
Comment (mikehalmamoj) [inline src/main/java/uk/gov/justice/hmpps/prison/api/model/Alias.java:64]: Add offenderId to aliases so we know which alias an identifier is associated with

File: src/main/java/uk/gov/justice/hmpps/prison/api/model/OffenderIdentifier.java [hunk 4]
```diff
@@ -45,17 +49,6 @@ public class OffenderIdentifier {
...
-        this.issuedAuthorityText = issuedAuthorityText;
-        this.issuedDate = issuedDate;
-        this.caseloadType = caseloadType;
-        this.whenCreated = whenCreated;
-    }
-
-    public OffenderIdentifier() {
-    }
+    @Schema(description = "Offender ID", example = "234547")
+    private Long offenderId;
```
Comment (mikehalmamoj) [inline src/main/java/uk/gov/justice/hmpps/prison/api/model/OffenderIdentifier.java:53]: Add offenderId to offenderIdentifier so we know if it relates to the main offender or an alias

File: src/main/java/uk/gov/justice/hmpps/prison/api/model/PrisonerSearchDetails.kt [hunk 5]
```diff
@@ -68,6 +71,9 @@ data class PrisonerSearchDetails(
   @Schema(description = "Prisoner Identifiers")
   val identifiers: List<OffenderIdentifier>? = null,
 
+  @Schema(description = "Prisoner Identifiers including those from aliases")
+  val allIdentifiers: List<OffenderIdentifier>? = null,
+
```
Comment (mikehalmamoj) [inline src/main/java/uk/gov/justice/hmpps/prison/api/model/PrisonerSearchDetails.kt:76]: Add `allIdentifiers` to the response object which will eventually replace `identifiers`

File: src/main/resources/db/migration/data/R__3_7__OFFENDER_IDENTIFIERS.sql [hunk 6]
```diff
@@ -15,3 +15,7 @@ INSERT INTO OFFENDER_IDENTIFIERS (OFFENDER_ID, OFFENDER_ID_SEQ, IDENTIFIER_TYPE,
...
 INSERT INTO OFFENDER_IDENTIFIERS (OFFENDER_ID, OFFENDER_ID_SEQ, IDENTIFIER_TYPE, IDENTIFIER, ISSUED_DATE, ROOT_OFFENDER_ID, CASELOAD_TYPE, VERIFIED_FLAG ) VALUES (-1003, 3, 'PNC', '24/1234567L', TO_DATE('2024-01-01', 'YYYY-MM-DD'), -1003, 'INST', 'Y');
+INSERT INTO OFFENDER_IDENTIFIERS (OFFENDER_ID, OFFENDER_ID_SEQ, IDENTIFIER_TYPE, IDENTIFIER, ISSUED_DATE, ROOT_OFFENDER_ID, CASELOAD_TYPE, VERIFIED_FLAG ) VALUES (-1012, 1, 'PNC', '2024/1234588L', TO_DATE('2024-01-01', 'YYYY-MM-DD'), -1012, 'INST', 'Y');
+INSERT INTO OFFENDER_IDENTIFIERS (OFFENDER_ID, OFFENDER_ID_SEQ, IDENTIFIER_TYPE, IDENTIFIER, ISSUED_DATE, ROOT_OFFENDER_ID, CASELOAD_TYPE, VERIFIED_FLAG ) VALUES (-1012, 2, 'NINO', 'DR784343E', TO_DATE('2024-01-01', 'YYYY-MM-DD'), -1012, 'INST', 'Y');
+INSERT INTO OFFENDER_IDENTIFIERS (OFFENDER_ID, OFFENDER_ID_SEQ, IDENTIFIER_TYPE, IDENTIFIER, ISSUED_DATE, ROOT_OFFENDER_ID, CASELOAD_TYPE, VERIFIED_FLAG ) VALUES (-1013, 1, 'PNC', '24/1234588L', TO_DATE('2024-01-01', 'YYYY-MM-DD'), -1012, 'INST', 'Y');
```
Comment (petergphillips) [inline src/main/resources/db/migration/data/R__3_7__OFFENDER_IDENTIFIERS.sql:20]: this is slightly confusing as the two PNCs are identical - one in short form and the other in long form.  Might be better to have different PNC values to be clearer what the intent is

File: src/main/java/uk/gov/justice/hmpps/prison/repository/jpa/model/OffenderIdentifier.java [hunk 7]
```diff
@@ -63,6 +63,9 @@ public static class OffenderIdentifierPK implements Serializable {
     @Column(name = "CASELOAD_TYPE")
     private String caseloadType;
 
+    @Column(name = "ISSUED_AUTHORITY_TEXT")
+    private String issuedAuthorityText;
```
Comment (petergphillips) [inline src/main/java/uk/gov/justice/hmpps/prison/repository/jpa/model/OffenderIdentifier.java:67]: is this always `INST`? Just wondering why we now need it

File: src/main/java/uk/gov/justice/hmpps/prison/api/model/OffenderIdentifier.java (hunk 4 above)
Comment (petergphillips) [reply src/main/java/uk/gov/justice/hmpps/prison/api/model/OffenderIdentifier.java:53]: would have been nicer to convert the file to kotlin 😁

File: src/main/resources/db/migration/data/R__3_7__OFFENDER_IDENTIFIERS.sql (hunk 6 above)
Comment (mikehalmamoj) [reply src/main/resources/db/migration/data/R__3_7__OFFENDER_IDENTIFIERS.sql:20]: It's just one of the scenarios I've seen in real world. Key thing is that all identifiers for all aliases are returned

File: src/main/java/uk/gov/justice/hmpps/prison/repository/jpa/model/OffenderIdentifier.java (hunk 7 above)
Comment (mikehalmamoj) [reply src/main/java/uk/gov/justice/hmpps/prison/repository/jpa/model/OffenderIdentifier.java:67]: This seems to contain free text entered by users - things like 'Prisoner has 2 PNCs' and 'This is the wrong PNC' wehre there are multiple different PNCs. So it seemed good to expose this text to clients.

**Review by petergphillips — APPROVED** [review]
LGTM

**Pull request merged**